File: backend/websocket_events.py
"""WebSocket event types and emission helpers for structured real-time communication."""

from typing import TypedDict, Literal, Union, Optional, Any
from datetime import datetime, timezone
import json


class BaseEvent(TypedDict):
    """Base structure for all WebSocket events."""
    type: str
    timestamp: str


class AgentStartedEvent(TypedDict):
    """Agent workflow started."""
    type: Literal["agent_started"]
    agent: str
    timestamp: str


class AgentCompletedEvent(TypedDict):
    """Agent workflow completed."""
    type: Literal["agent_completed"]
    agent: str
    summary: str
    timestamp: str


class ActivityEvent(TypedDict):
    """General activity/progress message."""
    type: Literal["activity"]
    message: str
    severity: Literal["info", "warn", "error", "success"]
    timestamp: str


class ConfidenceUpdateEvent(TypedDict):
    """Confidence score update."""
    type: Literal["confidence_update"]
    value: float
    timestamp: str


class TimelineEventType(TypedDict):
    """Operational timeline event."""
    type: Literal["timeline_event"]
    title: str
    time: str
    timestamp: str


class IntegrationEvent(TypedDict):
    """Integration/side-effect event (GitHub, Slack, etc.)."""
    type: Literal["integration"]
    service: str
    status: Literal["pending", "in_progress", "completed", "failed"]
    message: str
    timestamp: str


class RootCauseUpdateEvent(TypedDict):
    """Root cause analysis update."""
    type: Literal["root_cause_update"]
    rootCause: str
    confidence: float
    timestamp: str


class RemediationUpdateEvent(TypedDict):
    """Remediation action update."""
    type: Literal["remediation_update"]
    action: str
    timestamp: str


# Discriminated union of all event types
WebSocketEvent = Union[
    AgentStartedEvent,
    AgentCompletedEvent,
    ActivityEvent,
    ConfidenceUpdateEvent,
    TimelineEventType,
    IntegrationEvent,
    RootCauseUpdateEvent,
    RemediationUpdateEvent,
]


def get_iso_timestamp() -> str:
    """Get current timestamp in ISO format."""
    return datetime.now(timezone.utc).isoformat()


def emit_agent_started(agent: str) -> AgentStartedEvent:
    """Emit agent workflow started event."""
    return {
        "type": "agent_started",
        "agent": agent,
        "timestamp": get_iso_timestamp(),
    }


def emit_agent_completed(agent: str, summary: str) -> AgentCompletedEvent:
    """Emit agent workflow completed event."""
    return {
        "type": "agent_completed",
        "agent": agent,
        "summary": summary,
        "timestamp": get_iso_timestamp(),
    }


def emit_activity(message: str, severity: str = "info") -> ActivityEvent:
    """Emit activity/progress message event."""
    return {
        "type": "activity",
        "message": message,
        "severity": severity,  # type: ignore
        "timestamp": get_iso_timestamp(),
    }


def emit_confidence_update(value: float) -> ConfidenceUpdateEvent:
    """Emit confidence score update event."""
    return {
        "type": "confidence_update",
        "value": max(0.0, min(1.0, value)),
        "timestamp": get_iso_timestamp(),
    }


def emit_timeline_event(title: str, time: str) -> TimelineEventType:
    """Emit operational timeline event."""
    return {
        "type": "timeline_event",
        "title": title,
        "time": time,
        "timestamp": get_iso_timestamp(),
    }


def emit_integration(service: str, status: str, message: str) -> IntegrationEvent:
    """Emit integration/side-effect event."""
    return {
        "type": "integration",
        "service": service,
        "status": status,  # type: ignore
        "message": message,
        "timestamp": get_iso_timestamp(),
    }


def emit_root_cause_update(root_cause: str, confidence: float) -> RootCauseUpdateEvent:
    """Emit root cause update event."""
    return {
        "type": "root_cause_update",
        "rootCause": root_cause,
        "confidence": max(0.0, min(1.0, confidence)),
        "timestamp": get_iso_timestamp(),
    }


def emit_remediation_update(action: str) -> RemediationUpdateEvent:
    """Emit remediation action update event."""
    return {
        "type": "remediation_update",
        "action": action,
        "timestamp": get_iso_timestamp(),
    }


def event_to_json(event: WebSocketEvent) -> str:
    """Convert event to JSON string for WebSocket transmission."""
    return json.dumps(event)

File: frontend/src/lib/websocket-events.ts
/**
 * WebSocket event type definitions for SentinelOps.
 * Structured event types enable type-safe, event-driven architecture.
 */

export type WebSocketEventType =
  | "agent_started"
  | "agent_completed"
  | "activity"
  | "confidence_update"
  | "timeline_event"
  | "integration"
  | "root_cause_update"
  | "remediation_update"
  | "planner_output"
  | "snapshot_event";

export type SnapshotEvent = {
  type: "snapshot_event";
  agent: string;
  status: 'stored' | 'failed' | 'recovered' | 'replayed';
  message: string;
  timestamp: string;
};

export type PlannerTask = {
  agent: string;
  objective: string;
};

export type PlannerOutputEvent = {
  type: "planner_output";
  incident_type: string;
  tasks: PlannerTask[];
  timestamp: string;
};

export type AgentStartedEvent = {
  type: "agent_started";
  agent: string;
  timestamp: string;
};

export type AgentCompletedEvent = {
  type: "agent_completed";
  agent: string;
  summary: string;
  timestamp: string;
};

export type ActivityEventMessage = {
  type: "activity";
  message: string;
  severity: "info" | "warn" | "error" | "success";
  timestamp: string;
};

export type ConfidenceUpdateEvent = {
  type: "confidence_update";
  value: number;
  timestamp: string;
};

export type TimelineEventMessage = {
  type: "timeline_event";
  title: string;
  time: string;
  timestamp: string;
};

export type IntegrationEventMessage = {
  type: "integration";
  service: string;
  status: "pending" | "in_progress" | "completed" | "failed";
  message: string;
  timestamp: string;
};

export type RootCauseUpdateEvent = {
  type: "root_cause_update";
  rootCause: string;
  confidence: number;
  timestamp: string;
};

export type RemediationUpdateEvent = {
  type: "remediation_update";
  action: string;
  remediation: string;
  remediation_command?: string;
  requires_approval?: boolean;
  timestamp: string;
};

// Discriminated union of all event types
export type WebSocketEvent =
  | AgentStartedEvent
  | AgentCompletedEvent
  | ActivityEventMessage
  | ConfidenceUpdateEvent
  | TimelineEventMessage
  | IntegrationEventMessage
  | RootCauseUpdateEvent
  | RemediationUpdateEvent
  | SnapshotEvent
  | PlannerOutputEvent
  | InterventionRequiredEvent
  | InterventionResolvedEvent;

export interface InterventionRequiredEvent {
  type: 'intervention_required';
  incident_id: string;
  agent: string;
  reason: string;
  confidence: number;
  timestamp: string;
}

export interface InterventionResolvedEvent {
  type: 'intervention_resolved';
  incident_id: string;
  status: 'approved' | 'rejected';
  agent: string;
  timestamp: string;
}

/**
 * Type guard functions for discriminating event types.
 */

export function isPlannerOutputEvent(event: unknown): event is PlannerOutputEvent {
  return typeof event === "object" && event !== null && (event as any).type === "planner_output";
}

export function isAgentStartedEvent(event: unknown): event is AgentStartedEvent {
  return typeof event === "object" && event !== null && (event as any).type === "agent_started";
}

export function isAgentCompletedEvent(event: unknown): event is AgentCompletedEvent {
  return typeof event === "object" && event !== null && (event as any).type === "agent_completed";
}

export function isActivityEvent(event: unknown): event is ActivityEventMessage {
  return typeof event === "object" && event !== null && (event as any).type === "activity";
}

export function isConfidenceUpdateEvent(event: unknown): event is ConfidenceUpdateEvent {
  return typeof event === "object" && event !== null && (event as any).type === "confidence_update";
}

export function isTimelineEvent(event: unknown): event is TimelineEventMessage {
  return typeof event === "object" && event !== null && (event as any).type === "timeline_event";
}

export function isIntegrationEvent(event: unknown): event is IntegrationEventMessage {
  return typeof event === "object" && event !== null && (event as any).type === "integration";
}

export function isRootCauseUpdateEvent(event: unknown): event is RootCauseUpdateEvent {
  return typeof event === "object" && event !== null && (event as any).type === "root_cause_update";
}

export function isRemediationUpdateEvent(event: unknown): event is RemediationUpdateEvent {
  return typeof event === "object" && event !== null && (event as any).type === "remediation_update";
}

export function isSnapshotEvent(event: unknown): event is SnapshotEvent {
  return typeof event === "object" && event !== null && (event as any).type === "snapshot_event";
}

export function isInterventionRequiredEvent(event: unknown): event is InterventionRequiredEvent {
  return typeof event === "object" && event !== null && (event as any).type === "intervention_required";
}

export function isInterventionResolvedEvent(event: unknown): event is InterventionResolvedEvent {
  return typeof event === "object" && event !== null && (event as any).type === "intervention_resolved";
}
